In [4]:
!nvidia-smi

Wed Oct  2 09:41:30 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA H100 PCIe               On  | 00000000:21:00.0 Off |                   On |
| N/A   55C    P0             220W / 310W |   9112MiB / 81559MiB |     N/A      Default |
|                                         |                      |              Enabled |
+-----------------------------------------+----------------------+--

In [2]:
import torch

In [ ]:
torch.cuda.is_available()

In [ ]:
import os
import multiprocessing

# Get the number of CPU cores
num_cores = os.cpu_count()  # Alternatively, you can use multiprocessing.cpu_count()
print(f"Number of CPU cores available: {num_cores}")

In [1]:
import os
from PIL import Image
import glob
import matplotlib.pyplot as plt

In [2]:
def resize_images(input_folder, output_folder, width, height, suffix=''):
    # Create the output directory if it doesn't exist
    os.makedirs(output_folder, exist_ok=True)
    
    # Loop through all files in the input folder
    for filename in os.listdir(input_folder):
        # Check if the file is an image
        if filename.lower().endswith(('.jpg', '.jpeg', '.png', '.gif', '.bmp')):
            # Open the image
            img = Image.open(os.path.join(input_folder, filename))
            
            # Resize the image
            resized_img = img.resize((width, height), resample=Image.BICUBIC)
            
            # Modify the filename only if a suffix is provided
            if suffix:
                name, ext = os.path.splitext(filename)
                new_filename = f"{name}{suffix}{ext}"
            else:
                new_filename = filename
            
            # Save the resized image to the output folder with the new filename
            resized_img.save(os.path.join(output_folder, new_filename))

    print("Image resizing completed.")

In [6]:
# Example usage
# input_folder_1 = "workspace/gkd_large/DC0/train/HR_512/HR"
# output_folder_1 = "workspace/gkd_large/DC0/train/HR_128/HR"
# size_1 = 128
# resize_images(input_folder_1, output_folder_1, size_1, size_1)
# input_folder_13 = "workspace/gkd_large/DC2/train/HR_512/HR"
# output_folder_13 = "workspace/gkd_large/DC2/train/HR_128/HR"
# size_13 = 128
# resize_images(input_folder_13, output_folder_13, size_13, size_13)
# input_folder_14 = "workspace/gkd_large/DC0/train/HR_64/HR"
# output_folder_14 = "workspace/gkd_large/DC0/train/HR_64/LR_bicubic/X8"
# size_14 = 8
# resize_images(input_folder_14, output_folder_14, size_14, size_14, suffix='x8')
input_folder_2 = "workspace/gkd_large/DC2/train/HR_128/HR"
output_folder_2 = "workspace/gkd_large/DC2/train/HR_128/LR_bicubic/X2"
size_2 = 64
resize_images(input_folder_2, output_folder_2, size_2, size_2, suffix='x2')
input_folder_15 = "workspace/gkd_large/DC2/train/HR_128/HR"
output_folder_15 = "workspace/gkd_large/DC2/train/HR_128/LR_bicubic/X4"
size_15 = 32
resize_images(input_folder_15, output_folder_15, size_15, size_15, suffix='x4')
input_folder_16 = "workspace/gkd_large/DC2/train/HR_128/HR"
output_folder_16 = "workspace/gkd_large/DC2/train/HR_128/LR_bicubic/X8"
size_16 = 16
resize_images(input_folder_16, output_folder_16, size_16, size_16, suffix='x8')
input_folder_3 = "workspace/gkd_large/DC2/test/HR_128/bad/HR"
output_folder_3 = "workspace/gkd_large/DC2/test/HR_128/bad/LR_bicubic/X2"
size_3 = 64
resize_images(input_folder_3, output_folder_3, size_3, size_3, suffix='x2')
input_folder_17 = "workspace/gkd_large/DC2/test/HR_128/bad/HR"
output_folder_17 = "workspace/gkd_large/DC2/test/HR_128/bad/LR_bicubic/X4"
size_17 = 32
resize_images(input_folder_17, output_folder_17, size_17, size_17, suffix='x4')
input_folder_18 = "workspace/gkd_large/DC2/test/HR_128/bad/HR"
output_folder_18 = "workspace/gkd_large/DC2/test/HR_128/bad/LR_bicubic/X8"
size_18 = 16
resize_images(input_folder_18, output_folder_18, size_18, size_18, suffix='x8')
input_folder_4 = "workspace/gkd_large/DC2/test/HR_128/good/HR"
output_folder_4 = "workspace/gkd_large/DC2/test/HR_128/good/LR_bicubic/X2"
size_4 = 64
resize_images(input_folder_4, output_folder_4, size_4, size_4, suffix='x2')
input_folder_19 = "workspace/gkd_large/DC2/test/HR_128/good/HR"
output_folder_19 = "workspace/gkd_large/DC2/test/HR_128/good/LR_bicubic/X4"
size_19 = 32
resize_images(input_folder_19, output_folder_19, size_19, size_19, suffix='x4')
input_folder_20 = "workspace/gkd_large/DC2/test/HR_128/good/HR"
output_folder_20 = "workspace/gkd_large/DC2/test/HR_128/good/LR_bicubic/X8"
size_20 = 16
resize_images(input_folder_20, output_folder_20, size_20, size_20, suffix='x8')
# input_folder_5 = "workspace/gkd_large/DC0/test/HR_64/bad/HR"
# output_folder_5 = "workspace/gkd_large/DC0/test/HR_64/bad/LR_bicubic/X2"
# size_5 = 32
# resize_images(input_folder_5, output_folder_5, size_5, size_5, suffix='x2')
# input_folder_21 = "workspace/gkd_large/DC0/test/HR_64/bad/HR"
# output_folder_21 = "workspace/gkd_large/DC0/test/HR_64/bad/LR_bicubic/X4"
# size_21 = 16
# resize_images(input_folder_21, output_folder_21, size_21, size_21, suffix='x4')
# input_folder_22 = "workspace/gkd_large/DC0/test/HR_64/bad/HR"
# output_folder_22 = "workspace/gkd_large/DC0/test/HR_64/bad/LR_bicubic/X8"
# size_22 = 8
# resize_images(input_folder_22, output_folder_22, size_22, size_22, suffix='x8')
# input_folder_6 = "workspace/gkd_large/DC0/test/HR_32/bad/HR"
# output_folder_6 = "workspace/gkd_large/DC0/test/HR_32/bad/LR_bicubic/X2"
# size_6 = 16
# resize_images(input_folder_6, output_folder_6, size_6, size_6, suffix='x2')
# input_folder_23 = "workspace/gkd_large/DC0/test/HR_32/bad/HR"
# output_folder_23 = "workspace/gkd_large/DC0/test/HR_32/bad/LR_bicubic/X4"
# size_23 = 8
# resize_images(input_folder_23, output_folder_23, size_23, size_23, suffix='x4')
# input_folder_24 = "workspace/gkd_large/DC0/test/HR_32/bad/HR"
# output_folder_24 = "workspace/gkd_large/DC0/test/HR_32/bad/LR_bicubic/X8"
# size_24 = 4
# resize_images(input_folder_24, output_folder_24, size_24, size_24, suffix='x8')
# input_folder_7 = "workspace/gkd_large/DC0/test/HR_64/good/HR"
# output_folder_7 = "workspace/gkd_large/DC0/test/HR_64/good/LR_bicubic/X2"
# size_7 = 32
# resize_images(input_folder_7, output_folder_7, size_7, size_7, suffix='x2')
# input_folder_25 = "workspace/gkd_large/DC0/test/HR_64/good/HR"
# output_folder_25 = "workspace/gkd_large/DC0/test/HR_64/good/LR_bicubic/X4"
# size_25 = 16
# resize_images(input_folder_25, output_folder_25, size_25, size_25, suffix='x4')
# input_folder_26 = "workspace/gkd_large/DC0/test/HR_64/good/HR"
# output_folder_26 = "workspace/gkd_large/DC0/test/HR_64/good/LR_bicubic/X8"
# size_26 = 8
# resize_images(input_folder_26, output_folder_26, size_26, size_26, suffix='x8')
# input_folder_8 = "workspace/gkd_large/DC0/test/HR_32/good/HR"
# output_folder_8 = "workspace/gkd_large/DC0/test/HR_32/good/LR_bicubic/X2"
# size_8 = 16
# resize_images(input_folder_8, output_folder_8, size_8, size_8, suffix='x2')
# input_folder_27 = "workspace/gkd_large/DC0/test/HR_32/good/HR"
# output_folder_27 = "workspace/gkd_large/DC0/test/HR_32/good/LR_bicubic/X4"
# size_27 = 8
# resize_images(input_folder_27, output_folder_27, size_27, size_27, suffix='x4')
# input_folder_28 = "workspace/gkd_large/DC0/test/HR_32/good/HR"
# output_folder_28 = "workspace/gkd_large/DC0/test/HR_32/good/LR_bicubic/X8"
# size_28 = 4
# resize_images(input_folder_28, output_folder_28, size_28, size_28, suffix='x8')
# input_folder_9 = "workspace/gkd_large/DC0/test/HR_64/bad/HR"
# output_folder_9 = "workspace/gkd_large/DC0/test/HR_64/bad/LR_bicubic/X2"
# size_9 = 32
# resize_images(input_folder_9, output_folder_9, size_9, size_9, suffix='x2')
# input_folder_29 = "workspace/gkd_large/DC0/test/HR_64/bad/HR"
# output_folder_29 = "workspace/gkd_large/DC0/test/HR_64/bad/LR_bicubic/X4"
# size_29 = 16 
# resize_images(input_folder_29, output_folder_29, size_29, size_29, suffix='x4')
# input_folder_30 = "workspace/gkd_large/DC0/test/HR_64/bad/HR"
# output_folder_30 = "workspace/gkd_large/DC0/test/HR_64/bad/LR_bicubic/X8"
# size_30 = 8
# resize_images(input_folder_30, output_folder_30, size_30, size_30, suffix='x8')
# input_folder_10 = "workspace/gkd_large/DC0/test/HR_32/bad/HR"
# output_folder_10 = "workspace/gkd_large/DC0/test/HR_32/bad/LR_bicubic/X2"
# size_10 = 16
# resize_images(input_folder_10, output_folder_10, size_10, size_10, suffix='x2')
# input_folder_31 = "workspace/gkd_large/DC0/test/HR_32/bad/HR"
# output_folder_31 = "workspace/gkd_large/DC0/test/HR_32/bad/LR_bicubic/X4"
# size_31 = 8
# resize_images(input_folder_31, output_folder_31, size_31, size_31, suffix='x4')
# input_folder_32 = "workspace/gkd_large/DC0/test/HR_32/bad/HR"
# output_folder_32 = "workspace/gkd_large/DC0/test/HR_32/bad/LR_bicubic/X8"
# size_32 = 4
# resize_images(input_folder_32, output_folder_32, size_32, size_32, suffix='x8')
# input_folder_11 = "workspace/gkd_large/DC2/test/HR_64/good/HR"
# output_folder_11 = "workspace/gkd_large/DC2/test/HR_64/good/LR_bicubic/X2"
# size_11 = 32
# resize_images(input_folder_11, output_folder_11, size_11, size_11, suffix='x2')
# input_folder_33 = "workspace/gkd_large/DC2/test/HR_64/good/HR"
# output_folder_33 = "workspace/gkd_large/DC2/test/HR_64/good/LR_bicubic/X4"
# size_33 = 16
# resize_images(input_folder_33, output_folder_33, size_33, size_33, suffix='x4')
# input_folder_34 = "workspace/gkd_large/DC2/test/HR_64/good/HR"
# output_folder_34 = "workspace/gkd_large/DC2/test/HR_64/good/LR_bicubic/X8"
# size_34 = 8
# resize_images(input_folder_34, output_folder_34, size_34, size_34, suffix='x8')
# input_folder_12 = "workspace/gkd_large/DC2/test/HR_32/good/HR"
# output_folder_12 = "workspace/gkd_large/DC2/test/HR_32/good/LR_bicubic/X2"
# size_12 = 16
# resize_images(input_folder_12, output_folder_12, size_12, size_12, suffix='x2')
# input_folder_35 = "workspace/gkd_large/DC2/test/HR_32/good/HR"
# output_folder_35 = "workspace/gkd_large/DC2/test/HR_32/good/LR_bicubic/X4"
# size_35 = 8
# resize_images(input_folder_35, output_folder_35, size_35, size_35, suffix='x4')
# input_folder_36 = "workspace/gkd_large/DC2/test/HR_32/good/HR"
# output_folder_36 = "workspace/gkd_large/DC2/test/HR_32/good/LR_bicubic/X8"
# size_36 = 4
# resize_images(input_folder_36, output_folder_36, size_36, size_36, suffix='x8')

Image resizing completed.
Image resizing completed.
Image resizing completed.
Image resizing completed.
Image resizing completed.
Image resizing completed.
Image resizing completed.
Image resizing completed.
Image resizing completed.


In [ ]:
def show_images(images, number):
    for png_file in images[:number]:
        try:
            # Open the PNG file
            with Image.open(png_file) as img:
                # Get the image size
                width, height = img.size
                
                # Display the image and its size
                plt.figure()
                plt.imshow(img)
                plt.title(f"Size: {width}x{height}")
                plt.axis('off')  # Hide axes
                plt.show()
                
                print(f"Displayed {png_file} with size {width}x{height}")
        except Exception as e:
            print(f"Error displaying {png_file}: {e}")

In [ ]:
pred_images_dir = "./workspace/mvtec_anomaly_detection/carpet/train/LR_bicubic/X4"
pred_images = glob.glob(os.path.join(pred_images_dir, '*.png'))
show_images(pred_images, 2)

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

In [ ]:
ori_img_path = './workspace/images/ori_carpet_000.png'
recon_img_path = './workspace/images/pred_carpet_cut_x4_1000_000.png'
ori_img = Image.open(ori_img_path)
recon_img = Image.open(recon_img_path)
I_original = np.array(ori_img)
I_reconstructed = np.array(recon_img)

In [ ]:
I_difference = np.abs(I_original - I_reconstructed)
# I_difference = np.abs(I_reconstructed - I_original)

In [ ]:
I_normalized = (I_difference - I_difference.min(axis=(0, 1), keepdims=True)) / (I_difference.max(axis=(0, 1), keepdims=True) - I_difference.min(axis=(0, 1), keepdims=True))

I_normalized_gray = np.mean(I_normalized, axis=2)
heatmap = plt.cm.jet(I_normalized_gray)[:, :, :3]

I_original_norm = (I_original - I_original.min(axis=(0, 1), keepdims=True)) / (I_original.max(axis=(0, 1), keepdims=True) - I_original.min(axis=(0, 1), keepdims=True))

alpha = 0.5
overlay = alpha * heatmap + (1 - alpha) * I_original_norm

In [ ]:
# Display the results
plt.figure(figsize=(14, 6))

# Original Image
plt.subplot(1, 2, 1)
plt.imshow(I_original)
plt.title('Original Image')
plt.axis('off')

# Anomaly Score Map
plt.subplot(1, 2, 2)
plt.imshow(I_normalized_gray, cmap='jet')
plt.colorbar()
plt.title('Anomaly Score Map')
plt.axis('off')

# Overlay of Heatmap on Original Image
# plt.subplot(1, 3, 3)
# plt.imshow(overlay)
# plt.title('Overlay of Heatmap on Original Image')
# plt.axis('off')

plt.tight_layout()
plt.show()

In [ ]:
from PIL import Image
import numpy as np
import matplotlib.pyplot as plt
from skimage.metrics import structural_similarity as ssim

In [ ]:
ori_img_path = './workspace/images/ori_carpet_000.png'
recon_img_path = './workspace/images/pred_carpet_cut_x4_1000_000.png'
ori_img = Image.open(ori_img_path).convert('RGB')
recon_img = Image.open(recon_img_path).convert('RGB')
I_original = np.array(ori_img)
I_reconstructed = np.array(recon_img)

In [ ]:
# Convert images to grayscale for SSIM computation
I_original_gray = np.mean(I_original, axis=2)
I_reconstructed_gray = np.mean(I_reconstructed, axis=2)
# Debugging: Check the shapes of the grayscale images
print(f"Original Gray Shape: {I_original_gray.shape}")
print(f"Reconstructed Gray Shape: {I_reconstructed_gray.shape}")

In [ ]:
# Compute SSIM between the original and reconstructed images
ssim_map, ssim_full = ssim(I_original_gray, I_reconstructed_gray, data_range=255, full=True)

# Debugging: Check the shape of the SSIM map
print(f"SSIM Map Shape: {ssim_full.shape}")

In [ ]:
# Invert SSIM map to highlight anomalies (1 - SSIM)
anomaly_score_map = 1 - ssim_full

In [ ]:
# Normalize the anomaly score map for visualization
min_val = np.min(anomaly_score_map)
max_val = np.max(anomaly_score_map)
# Check if the min and max values are the same to avoid division by zero
if min_val != max_val:
    anomaly_score_map_normalized = (anomaly_score_map - min_val) / (max_val - min_val)
else:
    anomaly_score_map_normalized = anomaly_score_map

In [ ]:
# Display the results
plt.figure(figsize=(14, 6))

# Original Image
plt.subplot(1, 2, 1)
plt.imshow(I_original)
plt.title('Original Image')
plt.axis('off')

# Anomaly Score Map using SSIM
plt.subplot(1, 2, 2)
plt.imshow(anomaly_score_map_normalized, cmap='jet')
plt.colorbar()
plt.title('Anomaly Score Map (SSIM)')
plt.axis('off')

plt.tight_layout()
plt.show()

In [ ]:
# Directory containing TIFF images
input_dir = './workspace/experiment/efficientad/anomaly_maps/mvtec_ad/carpet/test/cut'
output_dir = './workspace/experiment/efficientad/anomaly_maps/mvtec_ad/carpet/test/cut'

In [ ]:
# Ensure the output directory exists
os.makedirs(output_dir, exist_ok=True)

In [ ]:
# Iterate over all TIFF files in the input directory
for filename in os.listdir(input_dir):
    if filename.endswith('.tiff'):
        # Open the TIFF image
        tiff_image = Image.open(os.path.join(input_dir, filename))
        
        # Convert to RGB (or 'L' for grayscale)
        tiff_image = tiff_image.convert('RGB')
        
        # Create the output filename
        output_filename = os.path.splitext(filename)[0] + '.png'
        
        # Save it as a PNG image
        tiff_image.save(os.path.join(output_dir, output_filename), 'PNG')

# GKD Dataset Preparation

In [1]:
import pandas as pd
from io import StringIO
df_dataset_description = pd.read_csv("./gkd-dataset/dataset-index.csv")
df_dataset_description[df_dataset_description['image_kind']==3]

filepath  image_kind  \
31360  NIO/2023_01_17_04_25_24_276_DC2_Gewebe_und_Nah...           3   
31361  NIO/2023_01_17_04_25_24_276_DC2_Gewebe_und_Nah...           3   
31362  NIO/2023_01_17_04_25_24_276_DC2_Gewebe_und_Nah...           3   
31363  NIO/2023_01_17_04_25_24_276_DC2_Gewebe_und_Nah...           3   
31364  NIO/2023_01_17_04_25_24_276_DC2_Gewebe_und_Nah...           3   
...                                                  ...         ...   
34715  NIO/2023_02_08_12_41_28_364_DC2_Gewebe_und_Nah...           3   
34716  NIO/2023_02_08_12_41_28_364_DC2_Gewebe_und_Nah...           3   
34717  NIO/2023_02_08_12_41_28_364_DC2_Gewebe_und_Nah...           3   
34718  NIO/2023_02_08_12_41_28_364_DC2_Gewebe_und_Nah...           3   
34719  NIO/2023_02_08_12_41_28_364_DC2_Gewebe_und_Nah...           3   

                                     original_image_name  x   y  
31360  2023_01_17_04_25_24_276_DC2_Gewebe_und_Naht_un...  0   0  
31361  2023_01_17_04_25_24_276_DC2_Gewebe_und_Naht_un...  0   1  
31362  2023_01_17_04_25_24_276_DC2_Gewebe_und_Naht_un...  0   2  
31363  2023_01_17_04_25_24_276_DC2_Gewebe_und_Naht_un...  0   3  
31364  2023_01_17_04_25_24_276_DC2_Gewebe_und_Naht_un...  0   4  
...                                                  ... ..  ..  
34715  2023_02_08_12_41_28_364_DC2_Gewebe_und_Naht_un...  0   9  
34716  2023_02_08_12_41_28_364_DC2_Gewebe_und_Naht_un...  0  10  
34717  2023_02_08_12_41_28_364_DC2_Gewebe_und_Naht_un...  0  11  
34718  2023_02_08_12_41_28_364_DC2_Gewebe_und_Naht_un...  0  12  
34719  2023_02_08_12_41_28_364_DC2_Gewebe_und_Naht_un...  0  13  

[3360 rows x 5 columns]

In [3]:
import numpy as np
import tensorflow as tf
from enum import Enum

class GkdImageSlice:
    def __init__(self, file_path, x: int, y: int, data: np.ndarray):
        self.file_path = file_path
        self.data = data
        self.x = x
        self.y = y

class GkdImageKind(Enum):
    UNLABELED_DC0=0
    NIO_DC0=1
    UNLABELED_DC2=2
    NIO_DC2=3
    IO_DC0=4
    IO_DC2=5

2024-10-10 09:38:26.844525: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-10-10 09:38:27.147921: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-10-10 09:38:27.234474: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-10-10 09:38:27.856048: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-10-10 09:38:29.816698: W tensorflow/compiler/tf2

In [4]:
class GkdDataSequence(tf.keras.utils.Sequence):
    def __init__(self, path, image_kind: GkdImageKind, batch_size, shuffle=True):
        super().__init__()
        self.path = path
        self.image_kind = image_kind
        
        df = pd.read_csv(self.path+"/dataset-index.csv")
        self.dataset_index = df[df["image_kind"]==self.image_kind.value].reset_index(drop=True)        
        #if self.image_kind == "unlabeled":
            #self.dataset_index =df[df["filepath"].str.startswith("unlabeled")].reset_index(drop=True)
        #elif self.image_kind == "NIO":
            #self.dataset_index =df[df["filepath"].str.startswith("NIO")].reset_index(drop=True)
            
        self.batch_size = batch_size
        self.shuffle=shuffle
        self.indexes = np.arange(len(self.dataset_index.index))
        self.on_epoch_end()
    
    def on_epoch_end(self):
        if self.shuffle:
            np.random.shuffle(self.indexes)
    
    def __len__(self):
        return int(np.floor(len(self.dataset_index.index)/self.batch_size))
    
    def __getitem__(self, batch_index):      
        # Generate data for the batch
        start_idx = batch_index * self.batch_size
        end_idx = (batch_index + 1) * self.batch_size
        batch_indexes = self.indexes[start_idx:end_idx]
        X, y = self.__data_generation(batch_indexes)

        """I am returning X two times here, since my example model
        is an Auto Encoder. In some cases this might not be what you want,
        so maybe try using y here after converting it to something usable
        """
        return np.asarray(X)/8191

    def __data_generation(self, batch_index):
        self.dataset_index.loc[batch_index]["filepath"]
        X = [np.fromfile("./gkd-dataset/"+path, dtype=np.float32).reshape(512, 512) for path in self.dataset_index.loc[batch_index]["filepath"]]
        y = self.dataset_index.loc[batch_index][["original_image_name", "x", "y"]].reset_index(drop=True)
        return X, y

# Save images as grayscale to desired folder

In [5]:
import os
from PIL import Image
import numpy as np

In [6]:
def save_img(image_array, output_path):
    # Ensure the image array is in the correct shape and data type
    if image_array.ndim == 3 and image_array.shape[0] == 1:
        # Squeeze the single channel to get the shape (height, width)
        grayscale_image = np.squeeze(image_array, axis=0)
    else:
        raise ValueError("Unexpected image array shape: {}".format(image_array.shape))
    # Normalize the image data to the range 0-255 and convert to uint8
    grayscale_image = ((grayscale_image - np.min(grayscale_image)) / (np.max(grayscale_image) - np.min(grayscale_image)) * 255).astype(np.uint8)
    # print(grayscale_image[0][0])

    grayscale_image = Image.fromarray(grayscale_image, mode='L')
    grayscale_image.save(output_path)

In [7]:
gkd_dataset_DC0_unlabeled = GkdDataSequence(path="./gkd-dataset", batch_size=1,image_kind=GkdImageKind.UNLABELED_DC0, shuffle=False)
gkd_dataset_DC0_IO = GkdDataSequence(path="./gkd-dataset", batch_size=1,image_kind=GkdImageKind.IO_DC0, shuffle=False)
gkd_dataset_DC0_NIO = GkdDataSequence(path="./gkd-dataset", batch_size=1,image_kind=GkdImageKind.NIO_DC0, shuffle=False)
gkd_dataset_DC2_unlabeled = GkdDataSequence(path="./gkd-dataset", batch_size=1,image_kind=GkdImageKind.UNLABELED_DC2, shuffle=False)
gkd_dataset_DC2_IO = GkdDataSequence(path="./gkd-dataset", batch_size=1,image_kind=GkdImageKind.IO_DC2, shuffle=False)
gkd_dataset_DC2_NIO = GkdDataSequence(path="./gkd-dataset", batch_size=1,image_kind=GkdImageKind.NIO_DC2, shuffle=False)

In [22]:
len(gkd_dataset_DC2_unlabeled)

14000

In [18]:
# Create the output directory if it doesn't exist
output_dir = []
output_dir.append('./workspace/gkd_large/DC0/train/HR_512/HR')
output_dir.append('./workspace/gkd_large/DC0/train/HR_128/HR')
output_dir.append('./workspace/gkd_large/DC0/train/HR_128/LR_bicubic/X2')
output_dir.append('./workspace/gkd_large/DC0/train/HR_128/LR_bicubic/X4')
output_dir.append('./workspace/gkd_large/DC0/train/HR_128/LR_bicubic/X8')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/good/HR')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/good/LR_bicubic/X2')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/good/LR_bicubic/X4')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/good/LR_bicubic/X8')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/bad/HR')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/bad/LR_bicubic/X2')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/bad/LR_bicubic/X4')
output_dir.append('./workspace/gkd_large/DC0/test/HR_128/bad/LR_bicubic/X8')
output_dir.append('./workspace/gkd_large/DC2/train/HR_512/HR')
output_dir.append('./workspace/gkd_large/DC2/train/HR_128/HR')
output_dir.append('./workspace/gkd_large/DC2/train/HR_128/LR_bicubic/X2')
output_dir.append('./workspace/gkd_large/DC2/train/HR_128/LR_bicubic/X4')
output_dir.append('./workspace/gkd_large/DC2/train/HR_128/LR_bicubic/X8')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/good/HR')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/good/LR_bicubic/X2')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/good/LR_bicubic/X4')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/good/LR_bicubic/X8')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/bad/HR')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/bad/LR_bicubic/X2')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/bad/LR_bicubic/X4')
output_dir.append('./workspace/gkd_large/DC2/test/HR_128/bad/LR_bicubic/X8')

In [19]:
for out_dir in output_dir:
    os.makedirs(out_dir, exist_ok=True)

In [ ]:
for i in range(448):
    print(i,gkd_dataset_DC2_IO.dataset_index.loc[i, "y"])

In [23]:
for i in range(14000):
    image_array_1 = gkd_dataset_DC0_unlabeled[i]
    original_position_1 = gkd_dataset_DC0_unlabeled.dataset_index.loc[i, "y"]
    original_name_1 = gkd_dataset_DC0_unlabeled.dataset_index.loc[i, "original_image_name"]
    output_path_1 = os.path.join("./workspace/gkd_large/DC0/train/HR_512/HR/", f'{i}_{original_position_1}.png')
    save_img(image_array_1, output_path_1)
    
    image_array_2 = gkd_dataset_DC2_unlabeled[i]
    original_position_2 = gkd_dataset_DC2_unlabeled.dataset_index.loc[i, "y"]
    original_name_2 = gkd_dataset_DC2_unlabeled.dataset_index.loc[i, "original_image_name"]
    output_path_2 = os.path.join("./workspace/gkd_large/DC2/train/HR_512/HR/", f'{i}_{original_position_2}.png')
    save_img(image_array_2, output_path_2)

unlabeled_dc2: 14000 images\
unlabeled_dc0: 6650 images\
nio_dc2: 3360 images\
nio_dc0: 1960 images\
io_dc0: 448 images\
io_dc2: 448 images

In [9]:
for i in range(3):
    image_array = gkd_dataset[i]
    print(image_array.ndim)
    print(image_array.shape[0])

NameError: name 'gkd_dataset' is not defined

In [ ]:
for i in range(3):
    image_array = gkd_dataset[i]
    print(f"Image {i} shape: {image_array.shape}, dtype: {image_array.dtype}")
    plt.imshow(image_array[0], cmap='gray')
    plt.show()

In [ ]:
import os
import shutil

# Source directory containing the 5 folders
source_dir = './workspace/mvtec_anomaly_detection_original/grid/test'

# Destination directory where all images will be moved
dest_dir = './workspace/mvtec_anomaly_detection_original/grid/bad'

# Create the destination directory if it doesn't exist
os.makedirs(dest_dir, exist_ok=True)

# Iterate through all subdirectories in the source directory
for folder_name in os.listdir(source_dir):
    folder_path = os.path.join(source_dir, folder_name)
    
    # Check if it's a directory
    if os.path.isdir(folder_path):
        # Iterate through all files in the folder
        for filename in os.listdir(folder_path):
            # Check if the file is a PNG image
            if filename.endswith('.png'):
                # Construct the new filename
                new_filename = f"{os.path.splitext(filename)[0]}_{folder_name}.png"
                
                # Construct full paths
                src_path = os.path.join(folder_path, filename)
                dest_path = os.path.join(dest_dir, new_filename)
                
                # Move and rename the file
                shutil.move(src_path, dest_path)

print("All images have been moved and renamed.")

In [1]:
import os
from PIL import Image
import numpy as np
from skimage import exposure

In [2]:
def min_max_scaling(image_array):
    min_val = np.min(image_array)
    max_val = np.max(image_array)
    return ((image_array - min_val) * (255 / (max_val - min_val))).astype(np.uint8)

In [3]:
def histogram_equalization(image_array):
    # Ensure the input array is in float format
    image_array = image_array.astype(float)
    
    # Normalize to 0-1 range
    image_array = (image_array - np.min(image_array)) / (np.max(image_array) - np.min(image_array))
    
    # Apply histogram equalization
    if len(image_array.shape) == 2:  # Grayscale image
        eq_array = exposure.equalize_hist(image_array)
    elif len(image_array.shape) == 3:  # Color image
        eq_array = np.dstack([exposure.equalize_hist(image_array[:,:,i]) 
                              for i in range(image_array.shape[2])])
    
    # Convert back to 0-255 range
    return (eq_array * 255).astype(np.uint8)

In [6]:
def process_images(source_folder, destination_folder):
    # Create destination folder if it doesn't exist
    if not os.path.exists(destination_folder):
        os.makedirs(destination_folder)

    # Process each PNG file in the source folder
    for filename in os.listdir(source_folder):
        if filename.lower().endswith('.png'):
            # Open the image
            img_path = os.path.join(source_folder, filename)
            with Image.open(img_path) as img:
                # Convert image to numpy array
                img_array = np.array(img)

                # Apply min-max scaling
                scaled_array = histogram_equalization(img_array)

                # Create a new image from the scaled array
                scaled_img = Image.fromarray(scaled_array)

                # Save the scaled image
                output_path = os.path.join(destination_folder, f"histo_{filename}")
                scaled_img.save(output_path)

            # print(f"Processed: {filename}")

    print("All images have been processed.")

In [8]:
mode = ['good', 'bad']
models = ['mvtec_carpet_512_X811:34:19_pretrained',
          'mvtec_carpet_512_X411:34:19_pretrained',
          'mvtec_grid_512_X411:34:19_pretrained', 
          'mvtec_grid_512_X811:34:19_pretrained', 
          'gkd_DC0_512_X411:34:19_pretrained',
          'gkd_DC0_512_X811:34:19_pretrained',
          'gkd_DC2_512_X411:34:19_pretrained',
          'gkd_DC2_512_X811:34:19_pretrained']

for model in models:
    for mod in mode:
        source_folder = f"workspace/images/drct/{model}/predicted_images/{mod}"
        destination_folder = f"workspace/images/drct/{model}/predicted_images/histo_after_{mod}"
        process_images(source_folder,destination_folder)

All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.
All images have been processed.


Processed: 018.png
Processed: 016.png
Processed: 008.png
Processed: 017.png
Processed: 007.png
Processed: 006.png
Processed: 000.png
Processed: 005.png
Processed: 025.png
Processed: 015.png
Processed: 019.png
Processed: 012.png
Processed: 010.png
Processed: 004.png
Processed: 021.png
Processed: 001.png
Processed: 023.png
Processed: 014.png
Processed: 022.png
Processed: 009.png
Processed: 026.png
Processed: 027.png
Processed: 002.png
Processed: 003.png
Processed: 024.png
Processed: 011.png
Processed: 013.png
Processed: 020.png
All images have been processed.


In [1]:
import numpy as np
from PIL import Image, ImageDraw, ImageFont

In [2]:
def create_pixel_value_image(image_path, output_path):
    # Open the image
    img = Image.open(image_path)
    
    # Convert the image to RGB mode if it's not already
    img = img.convert('RGB')
    
    # Get the dimensions of the image
    width, height = img.size
    
    # Create a new white image with the same dimensions
    result = Image.new('RGB', (width * 10, height * 10), color='white')
    draw = ImageDraw.Draw(result)
    
    # Try to load a font, fall back to default if not found
    try:
        font = ImageFont.truetype("arial.ttf", 9)  # Adjust size as needed
    except IOError:
        font = ImageFont.load_default()

    # Iterate through each pixel
    for y in range(height):
        for x in range(width):
            r, g, b = img.getpixel((x, y))
            avg = (r + g + b) // 3  # Average of RGB values
            
            # Draw the pixel value on the new image
            draw.text((x * 10, y * 10), str(avg), fill='black', font=font)

    # Save the result
    result.save(output_path)

In [3]:
input_image_path = 'workspace/images/drct/mvtec_carpet_32_X411:34:19_pretrained/predicted_images/bad/000_cut.png'
output_image_path = '000_cut.png'
create_pixel_value_image(input_image_path, output_image_path)

In [1]:
import os
import shutil

In [2]:
def create_directories(base_path):
    for quality in ['good', 'bad']:
        hr_path = os.path.join(base_path, quality, 'HR')
        lr_path = os.path.join(base_path, quality, 'LR_bicubic')
        os.makedirs(hr_path, exist_ok=True)
        os.makedirs(os.path.join(lr_path, 'X2'), exist_ok=True)
        os.makedirs(os.path.join(lr_path, 'X4'), exist_ok=True)

In [3]:
def copy_and_rename_files(base_path):
    for quality in ['good', 'bad']:
        source_path = os.path.join(base_path, quality)
        hr_path = os.path.join(source_path, 'HR')
        x2_path = os.path.join(source_path, 'LR_bicubic', 'X2')
        x4_path = os.path.join(source_path, 'LR_bicubic', 'X4')

        for filename in os.listdir(source_path):
            if filename.lower().endswith('.png'):
                # Copy to HR
                shutil.copy2(os.path.join(source_path, filename), hr_path)

                # Copy and rename for X2
                x2_filename = os.path.splitext(filename)[0] + 'x2.png'
                shutil.copy2(os.path.join(source_path, filename), os.path.join(x2_path, x2_filename))

                # Copy and rename for X4
                x4_filename = os.path.splitext(filename)[0] + 'x4.png'
                shutil.copy2(os.path.join(source_path, filename), os.path.join(x4_path, x4_filename))

In [4]:
def process_folders(folder_list):
    for folder in folder_list:
        create_directories(folder)
        copy_and_rename_files(folder)

In [5]:
folder_list = ['workspace_europa/images/drct/mvtec_carpet_128_X209:10:37/predicted_images',
               'workspace_europa/images/drct/mvtec_grid_128_X218:54:11/predicted_images',
               'workspace_europa/images/drct/gkd_DC0_128_X204:27:42/predicted_images',
               'workspace_europa/images/drct/gkd_DC2_128_X220:23:16/predicted_images',
              ]  
process_folders(folder_list)